File: Algorithms/338. Counting Bits/Solution.py
#coding=utf-8
__author__ = 'xuxuan'
import math
class Solution(object):
    _nums=[0]
    def countBits(self, num):
        """
        :type num: int
        :rtype: List[int]
        """
        if num<len(self._nums):
            return self._nums[:num+1]
        for i in range(len(self._nums),num+1):
            self._nums.append(self._nums[i-int(math.pow(2,int(math.log(i,2))))]+1)
        return self._nums

File: Algorithms/338. Counting Bits/Solution.java
public class Solution {
    List<Integer> list=new ArrayList<Integer>(){
        {
            add(0);
        }
    };
    public int[] countBits(int num) {
        if (num>=list.size()) {
            for (int i = list.size(); i < num+1; i++) {
                list.add(1+list.get(i-(int)Math.pow(2, (int)(Math.log(i)/Math.log(2)))));
            }
        }
        int[] ans=new int[num+1];
        for (int i = 0; i < ans.length; i++) {
            ans[i]=list.get(i);
        }
        return ans;
    }
}